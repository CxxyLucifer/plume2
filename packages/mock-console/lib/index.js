"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var MockConsole = /** @class */ (function () {
    function MockConsole() {
        this._logs = [];
        this.mock();
    }
    MockConsole.prototype.mock = function () { };
    MockConsole.prototype.mockFn = function () {
        var _this = this;
        return jest.fn(function (log) { return _this._logs.push(log); });
    };
    Object.defineProperty(MockConsole.prototype, "logs", {
        get: function () {
            return this._logs;
        },
        enumerable: true,
        configurable: true
    });
    MockConsole.prototype.reset = function () {
        this._logs = [];
    };
    return MockConsole;
}());
exports.MockConsole = MockConsole;
var MockWarn = /** @class */ (function (_super) {
    __extends(MockWarn, _super);
    function MockWarn() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MockWarn.prototype.mock = function () {
        console.warn = this.mockFn();
    };
    return MockWarn;
}(MockConsole));
exports.MockWarn = MockWarn;
var MockLog = /** @class */ (function (_super) {
    __extends(MockLog, _super);
    function MockLog() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MockLog.prototype.mock = function () {
        console.log = this.mockFn();
    };
    return MockLog;
}(MockConsole));
exports.MockLog = MockLog;
