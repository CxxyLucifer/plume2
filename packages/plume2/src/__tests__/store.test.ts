import { MockLog } from 'mock-console';
import { Action, Actor, IMap, Store } from '../index';

class HelloActor extends Actor {
  defaultState() {
    return { name: 'plume' };
  }

  @Action('change')
  change(state: IMap) {
    return state.set('name', 'plume++');
  }

  @Action('changeText')
  changeTexst(state: IMap, text) {
    return state.set('name', text);
  }
}

class LoadingActor extends Actor {
  defaultState() {
    return { loading: false };
  }

  @Action('change')
  change(state: IMap) {
    return state.set('loading', true);
  }
}

class AppStore extends Store {
  bindActor() {
    return [
      //binding actor
      HelloActor,
      LoadingActor
    ];
  }

  change = () => {
    this.dispatch('change');
  };

  changeTransation() {
    this.transaction(() => {
      this.dispatch('changeText', 'iflux2');
      this.dispatch('changeText', 'plume');
      this.dispatch('changeText', 'plume2');
    });
  }
}

describe('store test suite', () => {
  it('default state', () => {
    const store = new AppStore() as any;

    //defautlState
    expect(store.state().toJS()).toEqual({
      loading: false,
      name: 'plume'
    });

    //actors
    expect(store._actors.length).toEqual(2);

    //actor'state
    const actorState = store._actorsState;
    expect([actorState[0].toJS(), actorState[1].toJS()]).toEqual([
      { name: 'plume' },
      { loading: false }
    ]);
  });

  it('store dispatch', () => {
    const store = new AppStore() as any;
    store.change();

    const storeState = store.state();
    expect(storeState.toJS()).toEqual({ loading: true, name: 'plume++' });

    const actorsState = store._actorsState;
    expect([actorsState[0].toJS(), actorsState[1].toJS()]).toEqual([
      { name: 'plume++' },
      { loading: true }
    ]);
  });

  it('store transation disptch', () => {
    const store = new AppStore();
    store.changeTransation();
    store.subscribe(state => {
      expect(state.toJS()).toEqual({
        loading: false,
        name: 'plume2'
      });
    });
  });

  it('store subscribe', () => {
    const store = new AppStore({}) as any;
    //@ts-ignore
    const _handleStoreChange = (state: IMap) => {
      expect(store.state()).toEqual({ loading: true, name: 'plume++' });
    };

    store.subscribe(_handleStoreChange);
    expect(store._callbacks.length).toEqual(1);

    store.unsubscribe(_handleStoreChange);
    expect(store._callbacks.length).toEqual(0);
  });

  it('dispatch false', () => {
    class LoadingActor extends Actor {
      defaultState() {
        return { loading: true };
      }

      @Action('loading:end')
      end(state: IMap, status: boolean) {
        return state.set('loading', status);
      }
    }

    class AppStore extends Store {
      bindActor() {
        return [LoadingActor];
      }
    }

    const mock = new MockLog();
    const store = new AppStore({ debug: true });
    store.dispatch('loading:end', false);

    expect(store.state().toJS()).toEqual({
      loading: false
    });

    expect(mock.logs).toMatchSnapshot();
  });
});
