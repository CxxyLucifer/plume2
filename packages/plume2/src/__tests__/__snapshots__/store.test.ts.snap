// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store test suite dispatch false 1`] = `
Array [
  "store dispatch => 'loading:end'",
  "params|> false",
  "LoadingActor receive => 'loading:end'",
]
`;
